nixpkgs/nixpkgs.bzl

```
@@ -158,6 +160,85 @@ def _nixpkgs_package_impl(repository_ctx):
+            line = line.split(sep=' ')
+
+            # Heuristic: a dependency is something which looks like:
+            # evaluating file FILE
+            # copied source FILE
+            if (line[0], line[1]) in [("evaluating", "file"), ("copied", "source")]:
+                # We ignore some files:
+                # - Anything in /nix/store, they are not explicit dependencies are are supposed to be immutable
+                # - Anything from .cache/bazel, only case I encountered was a local nixpkgs clone handled by bazel
+                # - .config/nixpkgs. user configuration should not impact the reproducibility of the build
+                if (
+                   not line[2].startswith("'/nix/store")
+                   and ".cache/bazel" not in line[2]
+                   and ".config/nixpkgs" not in line[2]
+                ):
+                    filename = line[2][1:-1] # trimming quotes
+
+                    # This filename can be either a file or a directory
+                    # this find command will list all the sub files of a potential directory
+                    find_result = _execute_or_fail(
+                        repository_ctx,
+                        [_executable_path(repository_ctx, "find"), filename, "-type", "f", "-print0"],
+                    )
+
+                    # filenames are separated by \0 to allow filenames with newlines
+                    for filename in find_result.stdout.rstrip("\0").split("\0"):
+                        deps.append(filename)
+
+        # declared deps contains all the implicit dependencies declared by the user
+        # starting by all the files in `nix_file_deps`
+        # realpath is used to match files listed by nix
+        declared_deps = [str(repository_ctx.path(f).realpath) for f in repository_ctx.attr.nix_file_deps]
+
+        # extend declared deps with the list of all repositories files
+        if repository_ctx.attr.nix_file:
+            declared_deps.append(str(repository_ctx.path(repository_ctx.attr.nix_file)))
+        for rep in repositories.keys():
+            declared_deps.append(str(repository_ctx.path(rep).realpath))
+
+        # Set substraction deps - declared_deps must be empty
+        # Note: we do not fail if some declared deps are not
+        # necessary, better safe than sorry, this won't affect
+        # reproducibility, and we are not sure that the current
+        # heuristic can find all the indirect dependencies
+
+        deps_minus_declared_deps = []
+        for dep in deps:
+            if dep not in declared_deps:
+                deps_minus_declared_deps.append(dep)
+
+        if deps_minus_declared_deps:
+            fail("""
+
+Non hermetic configuration for repository {repo_name}.
+
+The following dependencies are not declared in *nixpkgs_package* attributes.
+
+You need to update the repository rule *{repo_name}* and set/extend *nix_file_deps* with the following values:
```

The list of suggested values contains absolute paths. Users will most likely not want to copy those verbatim, but will have to convert them to Bazel targets as appropriate for their workspace.
```suggestion
You need to update the repository rule *{repo_name}* and set/extend *nix_file_deps* with the following dependencies:
```